src/ansi-c/c_typecheck_expr.cpp

```diff
@@ -1945,6 +1949,73 @@ void c_typecheck_baset::typecheck_side_effect_function_call(
       warning().source_location=f_op.find_source_location();
       warning() << "function `" << identifier << "' is not declared" << eom;
     }
+    else if(
+      sym_entry->second.type.get_bool(ID_C_inlined) &&
+      sym_entry->second.is_macro && sym_entry->second.value.is_not_nil())
+    {
+      // calling a function marked as always_inline
+      const symbolt &func_sym = sym_entry->second;
+      const code_typet &func_type = to_code_type(func_sym.type);
+
+      replace_symbolt replace;
+
+      const code_typet::parameterst &parameters = func_type.parameters();
+      auto p_it = parameters.begin();
+      for(const auto &arg : expr.arguments())
+      {
+        if(p_it == parameters.end())
+        {
+          err_location(f_op);
+          error() << "function call has additional arguments, "
+                  << "cannot apply always_inline" << eom;
+          throw 0;
+        }
+
+        irep_idt p_id = p_it->get_identifier();
+        if(p_id.empty())
+        {
+          p_id = id2string(func_sym.base_name) + "::" +
+                 id2string(p_it->get_base_name());
+        }
+        replace.insert(p_id, arg);
+
+        ++p_it;
+      }
+
+      if(p_it != parameters.end())
+      {
+        err_location(f_op);
+        error() << "function call has missing arguments, "
+                << "cannot apply always_inline" << eom;
+        throw 0;
+      }
+
+      codet body = to_code(func_sym.value);
+      replace(body);
```

Comment:
So this will replace formal parameters with actuals, but not actually cause locals to alias because they retain the qualified names (`f::1::x` vs. `g::1::x`) they had originally? That seems right for inlining but wrong for macros?

Reply:
As said above, we don't actually do preprocessor macros in this way, so it should be fine.